apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -81,9 +81,18 @@ public enum Reply
     /// </summary>
     public static TimeSpan MinimumAckExtensionWindow { get; } = TimeSpan.FromMilliseconds(50);
 
+    /// <summary>
+    /// The default minimum message Acknowledgement deadline extension of 60 seconds for exactly-once delivery subscriptions.
```

Comment:
Is it default or is it minimum?
From all the internal emails, I understand it's minimum not default.

What the implications are on failing or replacing when user provided values are outside the range, is not clear to me, and I'm ultimately fine with either.

Note SubscriberClient.Settings has a Validate method that is not taking eod restrictions into account. 
That might be fine, as Settings seems to not know whether eod or non-eod, but then, we need to make those validations or replacements elsewhere.

Reply:
Please share your thoughts and feedback on the following:
1) I have kept 2 separate AckDeadlines for regular and eod subscriptions. (With the current defaults, both will be same but wanted to get the thoughts if we should keep two AckDeadlines or just one). Due to the defaults, the same variable will work for both the subscriptions, even if subscription is changed in-flight. 
2) I have added a new MinimumAckDeadlineForExactlyOnceDelivery and marked the DefaultMinimumAckExtensionWindowForExactlyOnceDelivery as Obsolete. Given we have take only AckDeadline and (not MinimumAckDeadline) would it make sense to create this or should be just use DefaultAckDeadline (as it has same default value as DefaultMinimumAckDeadline) and mark DefaultMinimumAckExtensionWindowForExactlyOnceDelivery as Obsolete. We can always change this if there is a requirement to change the defaults.

The requirement to support Minimum/Maximum AckDeadline is only for libraries that don't allow users to specify AckDeadline. In those libraries, the AckDeadline is calculated as a percentile value based on latency and then to ensure that the AckDeadline is within range, they make use of Maximum and MinimumAckDeadline values. In our library, we already allow user to set AckDeadline which is within a range (10,600), so we don't need this.

Reply:
The tldr is

I'm totally fine with using two variables for values that have different meaning even if currently they have the same value. My comments were not about that.

My comments were about the 60 seconds minimum for EOD ack deadline. You are not using that as a minimum, you are using it as a default. In the internal emails it says clearly that should be used as a minimum. Something like "even though the C# library does not have a heuristic to calculate the deadline extension, an thus it doesn't need a min and max to anchor the heuristic result, the deadline extension is constrained by the overall EOD min extension of 60 seconds".

I'm quoting that from memory but that's effectively what the Pub/Sub team said on the last email thread that you started.

The other thing I want to know is, where is the requirement of "not restarting the subscriber if the subscription changes from EOD to non EOD or viceversa" is coming from. Is that a requirement that's explicitly coming from the Pub/Sub team?

Reply:
> My comments were about the 60 seconds minimum for EOD ack deadline. You are not using that as a minimum, you are using it as a default. In the internal emails it says clearly that should be used as a minimum. Something like "even though the C# library does not have a heuristic to calculate the deadline extension, an thus it doesn't need a min and max to anchor the heuristic result, the deadline extension is constrained by the overall EOD min extension of 60 seconds".

Right Amanda. This is there in the email summary. However, this is not complete. This is the default minimum if user hasn't specified the minimumAckDeadline value. User can provide any value in range (10, 600) and override this default minimum. Since we do not have MinimumAckDeadline, so either we use the DefaultAckDeadline (which is 60 seconds) or we change it to DefaultMinimumAckDeadlineForExactlyOnceDelivery and use that. - What would be your preference?  (I will also paste the requirement in chat space for clarity)

>The other thing I want to know is, where is the requirement of "not restarting the subscriber if the subscription changes from EOD to non EOD or viceversa" is coming from. Is that a requirement that's explicitly coming from the Pub/Sub team?

Yes, this came as a bug logged by PubSub team. They explicitly stated to change the subscription in flight with subscriber on. I will share this in our chat as well. Thanks.

Please let me know your thoughts and I will update as per your suggestion/preference. Thanks.

Reply:
Thanks Jon for the detailed response.

The requirement is to have default minimum AckDeadline of 60 seconds which can be overridden by the user to any value in the valid range. However, this requirement is for libraries that support minimum/maximum AckDeadline. So, I am happy naming and using the default value as DefaultAckDeadlineForExactlyOnceDelivery as suggested above. I think we can do with just one AckDeadline property in user settings. 

 The reason for using two variables for modifyAckDeadline was that we get to know if subscription is eod or not only later when we get StreamingPullResponse that sets the flag if eod is enabled.

But yes with the above code everything is achievable. Please let me know if this sounds fine and I can implement that. Thank you.

Reply:
> I think we can do with just one AckDeadline property in user settings.

So we don't need to allow users to configure different deadlines based on whether it turns out that it's EOD or not? If that's the case, that's great.